docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,179 @@
+
+* Use `camelCase` to name non-constant local variables.
 
-A basic rule of formatting is to use "Visual Studio defaults".
-Here are some general guidelines
+* Use `PascalCase` to name constant local variables and fields.
+  The only exception is for interop code where the constant should exactly match the name and value of the code you are calling via interop (i.e. `const int ERROR_SUCCESS = 0`).
 
-* No tabs, indent 4 spaces.
-* Braces usually go on their own line,
+* Use `PascalCase` to name types and all other type members.
+
+### Layout Conventions
+
+* Use four spaces of indentation (no tabs).
+
+* Avoid more than one blank empty line at any time.
+
+* Avoid unnecessary trailing spaces at the end of a line.
+
+* Braces usually go on their own lines,
   with the exception of single line statements that are properly indented.
-* Use `_camelCase` for instance fields,
-  use `readonly` where possible.
+
+* Namespace imports should be specified at the top of the file,
+  outside of `namespace` declarations and should be sorted alphabetically.
+
+* Fields should be specified at the top within type declarations.
+
+* File encoding should be `ASCII`.
+  All `BOM` encodings should be avoided.
+  Tests that need a `BOM` encoding file should generate the file on the fly.
+
+### Member Conventions
+
 * Use of `this` is neither encouraged nor discouraged.
-* Avoid more than one blank empty line.
-* Public members should use [doc comments](https://msdn.microsoft.com/en-us/library/b2s063f7.aspx),
-  internal members may use doc comments but it is not encouraged.
+
+* Use `nameof(<member-name>)` instead of `"<member-name>"` whenever possible and relevant.
+
+* Always specify the visibility, even if it's the default (i.e. `private string _foo` not `string _foo`).
+  Visibility should be the first modifier (i.e. `public abstract` not `abstract public`).
+
+* Make members private where possible.
+  Avoid declaring public members unless it's absolutely necessary.
+
 * Public members in a namespace that ends with `Internal`,
   for example `System.Management.Automation.Internal` are not considered a supported public API.
   Such members are necessarily public as implementation details in code shared between C# and PowerShell script,
   or must be available publicly by generated code.
-* File encoding should be ASCII (preferred)
-  or UTF8 (with BOM) if absolutely necessary.
-
-## Preprocessor defines
 
-There are 3 primary preprocessor macros we define during builds:
+### Commenting Conventions
 
-* DEBUG - guard code that should not be included in release builds
-* CORECLR - guard code that differs between Full CLR and CoreCLR
-* UNIX - guard code that is specific to Unix (Linux and macOS)
+* Add comments when changes are not trivial or could be confusing.
 
-Any other preprocessor defines found in the source are used for one-off custom builds,
-typically to help debug specific scenarios.
+* Add comments when a reviewer needs help to understand your changes.
```

Comment:
I think phrasing can be better here. It may be perceived as encouragement for creating a historical log in the comments, i.e.

```
// This line used to be bla-bla 
//  and was changed to bla-bla because previously we had a problem in bla-bla corner case.
```

Reply:
The sentences are reworded to:
```
* Add comments where the code is not trivial or could be confusing.
* Add comments where a reviewer needs help to understand the code.
```